libraries/botbuilder-ai/src/qnaMakerDialog.ts

```diff
@@ -0,0 +1,359 @@
+    /**
+    * Options for QnAMaker knowledgebase.
+    */
+    qnaMakerOptions: QnAMakerOptions;
+    /**
+    * QnAMakerDialog response options.
+    */
+    qnaDialogResponseOptions: QnAMakerDialogResponseOptions;
+}
+
+/**
+ * Query QnAMaker knowledgebase using user utterance. The dialog will also present user with appropriate multi-turn prompt or active learning options.
+ */
+export class QnAMakerDialog extends WaterfallDialog {
+    
+    // state and step value key constants
+    private qnAContextData: string = 'previousContextData';
+    private previousQnAId: string = 'previousQnAId';
+    private options: string = 'options';
+    private qnAData: string = 'qnaData';
+    private currentQuery: string = 'currentQuery';
+
+    // Dialog options parameters
+    private defaultCardNoMatchResponse: string = "Thanks for the feedback.";
+    private defaultNoAnswer: string = "No QnAMaker answers found.";
+    private maximumScoreForLowScoreVariation: number = 0.95;
+
+    private knowledgeBaseId: any;
+    private hostName: any;
+    private endpointKey: any;
+    private threshold: number;
+    private top: number;
+    private noAnswer: Activity;
+    private activeLearningCardTitle: string;
+    private cardNoMatchText: string;
+    private strintFilters: any;
+    private cardNoMatchResponse: Activity;
+
+    /**
+     * Creates a new QnAMakerDialog instance.
+     * @param knowledgeBaseId The Id of the QnAMaker knowledgebase to be queried.
+     * @param endpointKey The endpoint key to use when querying the knowledgebase.
+     * @param hostName Hostname to be used to form the QnAMaker host URL, which follows the following format https://{hostName}.azurewebsites.net/qnamaker
+     * @param noAnswer (Optional) Activity to be sent in the event no answer is found within the knowledgebase.
+     * @param threshold (Optional) The threshold above which to treat answers found from the knowledgebase as a match.
+     * @param activeLearningCardTitle (Optional) Title of the card displayed showing active learning options if active learning is enabled.
+     * @param cardNoMatchText (Optional) Text to be show on a button alongside active learning options, allowing a user to indicate none of the options are applicable.
+     * @param top (Optional) Maximum number of answers to return from the knowledgebase.
+     * @param cardNoMatchResponse (Optional) Activity to be sent if the user selects the no match option on an active learning card.
+     * @param strictFilters (Optional) QnAMakerMetadata collection used to filter / boost queries to the knowledgebase.
+     * @param dialogId (Optional) Id of the created dialog. Default is 'QnAMakerDialog'.
+     */
+    constructor(knowledgeBaseId, endpointKey, hostName, noAnswer = null, threshold = 0.3, activeLearningCardTitle = 'Did you mean:', cardNoMatchText = 'None of the above.', top = 3, cardNoMatchResponse: Activity = null, strictFilters: QnAMakerMetadata[] = null, dialogId = 'QnAMakerDialog') {
```

Comment:
>null [](start = 67, length = 4)

Never use "null" in JS. Use either "undefined" or the "?" operator.

Reply:
Done.